pkg/http/auth.go

```diff
@@ -0,0 +1,215 @@
+	HandleChallenge(context.Context, *http.Response) error
+	// AuthorizeRequest is responsible for authorizing the request to be
+	// sent to the origin server.
+	AuthorizeRequest(context.Context, *http.Request) (*http.Request, error)
+}
+
+// AuthPolicy defines an authentication policy. It takes a response
+// and determines whether or not it warrants authentication.
+type AuthPolicy func(*http.Response) bool
+
+// DefaultAuthPolicy defines the default AuthPolicy, where by only a "401
+// Unauthorized" warrants authentication.
+var DefaultAuthPolicy = func(resp *http.Response) bool {
+	return resp.StatusCode == http.StatusUnauthorized
+}
+
+// AuthReqContextFunc takes the original request context and
+// returns a context that will be used by the new authenticated
+// request.
+type AuthReqContextFunc func(reqCtx context.Context) context.Context
+
+// DefaultAuthReqContext is the default AuthReqContextFunc. It returns
+// an entirely new context.
+var DefaultAuthReqContext = func(reqCtx context.Context) context.Context {
+	return context.Background()
+}
+
+// AuthClient provides a HTTP client that is capable of authenticating
+// with origin servers. It contains an AuthHandler type that is responsible
+// for preparing valid responses/answers to challenges as well authenticating
+// requests. It wraps an inner retryable client, that is uses to send requests.
+//
+// Note: The AuthClient does not directly provide a mechanism for caching
+// credentials/tokens. Ideally, this should be handled by the underlying
+// AuthHandler.
+type AuthClient struct {
+	client     *rhttp.Client
+	handler    AuthHandler
+	policy     AuthPolicy
+	getAuthCtx AuthReqContextFunc
+	headers    http.Header
+}
+
+type AuthClientOpt func(*AuthClient)
+
+// WithHeader adds a http.Header to the AuthClient that will
+// be attached to every request.
+func WithHeader(headers http.Header) AuthClientOpt {
+	return func(ac *AuthClient) {
+		ac.headers = headers
+	}
+}
+
+// WithAuthPolicy attaches an AuthPolicy to the AuthClient.
+func WithAuthPolicy(policy AuthPolicy) AuthClientOpt {
+	return func(ac *AuthClient) {
+		ac.policy = policy
+	}
+}
+
+// WithRetryableClient attaches a retryable client to the AuthClient.
+func WithRetryableClient(client *rhttp.Client) AuthClientOpt {
+	return func(ac *AuthClient) {
+		ac.client = client
+	}
+}
+
+// WithRetryableClient attaches a AuthReqContextFunc to the AuthClient.
+func WithAuthRequestCtxFunc(arc AuthReqContextFunc) AuthClientOpt {
+	return func(ac *AuthClient) {
+		ac.getAuthCtx = arc
+	}
+}
+
+// NewAuthClient creates a new AuthClient given an AuthHandler.
+//
+// An AuthHandler must be provided. If no retryable client is provided
+// it will create a default retryable client. If no AuthPolicy is
+// provided the DefaultAuthPolicy will be used. If no AuthReqCtxFunc
+// is provided the DefaultAuthReqContext is used.
+func NewAuthClient(authHandler AuthHandler, opts ...AuthClientOpt) (*AuthClient, error) {
```

Comment:
Because of the global client, `NewAuthClient` should only be invoked once, correct?

Context: Curious if the discussions around the options pattern being less performant than a configuration struct would apply here. Perhaps not if the invoke is just one time.

Reply:
Yeah, I have some conflicting thoughts on this. On one hand I think functional options have the benefit of making it easier to initialize a "default" instance, but on the other hand it makes initializing more complex instances harder to read. I think I'm going to keep it as is for now, but it might be worth revisiting in the future. I've made the `http` package internal, so we can make changes to the API freely without having to worry about external consumers.